core/market/decentralized/src/negotiation/errors.rs

```diff
@@ -31,13 +29,3 @@ pub enum QueryEventsError {
 
 #[derive(Error, Debug)]
 pub enum ProposalError {}
-
-impl From<TakeEventsError> for QueryEventsError {
-    fn from(e: TakeEventsError) -> Self {
-        match e {
-            TakeEventsError::SubscriptionExpired(id) => QueryEventsError::SubscriptionExpired(id),
-            TakeEventsError::SubscriptionNotFound(id) => QueryEventsError::Unsubscribed(id),
-            _ => QueryEventsError::FailedGetEvents(e),
-        }
-    }
-}
```

Comment:
I've changed it because it has flaw. If we've got `TakeEventsError::SubscriptionNotFound` we cannot translate it into `QueryEventsError::Unsubscribed` because it is not only possibility. Subscription might never existed. This change is not only for code brevity, but also for correctness

Reply:
Than you should rather change QueryEventsError to `SubscriptionNotFound`, because we can never return `QueryEventsError::Unsubscribed`

Reply:
I don't like this pattern of embedding errors in errors in every place possible. It makes sense in some places, but in some doesn't. QueryEventsError should give very specific errors without need to dispatch nesting.

Reply:
> Than you should rather change QueryEventsError to `SubscriptionNotFound`, because we can never return `QueryEventsError::Unsubscribed`
yes, we can and it is illustrated two times in test


> I don't like this pattern of embedding errors in errors in every place possible. It makes sense in some places, but in some doesn't. QueryEventsError should give very specific errors without need to dispatch nesting.

Here it seems to me it is one of the case when embedding works very well, because otherwise you need to repeat all patterns from `TakeEventsError` in `QueryEventsError`, isn't it?

Reply:
Repeating isn't a problem for me. Ugly api is a problem.

And what are patterns on which caller wants to match against here?
_NotFound_, _Unsubscribed_, _Expired_ are fundamental errors and caller can react somehow to them.
Other errors aren't so important, because caller could threat them as "some internal error" and he can't do anything more with this. He can just print message and wait for human to read it ;)

In this code the most important errors are hidden behind strange enum `TakeEventsError`. `TakeEventsError` tells him nothing, because he already knows that querying events failed. So he must dive deeper, to know what happened.

Embedding makes sense, when we deal with errors from internal layers. We want to wrap them with higher level errors, to add more meaningful message. When there are multiple layers, we have full path of errors that caused problem.
In such case wrapping adds information. Here we take information.

Reply:
I think we could leave it, but maybe we should make some meeting to establish common error handling practices, that would satisfy us all